File: iSistemiumCore/DataModel/STMFmdb.m
//
//  STMFMDB_OBJC.m
//  iSisSales
//
//  Created by Edgar Jan Vuicik on 16/12/2016.
//  Copyright © 2016 Sistemium UAB. All rights reserved.
//

//Note: The calls to FMDatabaseQueue's methods are blocking. So even though you are passing along blocks, they will not be run on another thread.

#import <Foundation/Foundation.h>
#import "STMFmdb.h"
#import "STMFunctions.h"
#import "FMDB.h"
#import "STMCoreObjectsController.h"
#import "STMPredicateToSQL.h"

@implementation STMFmdb

NSDictionary* columnsByTable;
FMDatabaseQueue *queue;
FMDatabasePool *pool;


- (instancetype)init {
    
    self = [super init];
    
    if (self) {
        
        NSArray *entityNames = STMCoreObjectsController.document.myManagedObjectModel.entitiesByName.allKeys;
        NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
        NSString *documentDirectory = [paths objectAtIndex:0];
        NSString *dbPath = [documentDirectory stringByAppendingPathComponent:@"database.db"];
        
        queue = [FMDatabaseQueue databaseQueueWithPath:dbPath];
        pool = [FMDatabasePool databasePoolWithPath:dbPath];
        
        NSMutableDictionary *columnsDictionary = @{}.mutableCopy;
        
        [queue inDatabase:^(FMDatabase *database){
            
            NSString *createIndexFormat = @"CREATE INDEX IF NOT EXISTS FK_%@_%@ on %@ (%@);";
            NSString *fkColFormat = @"%@ TEXT REFERENCES %@(id) ON DELETE %@";
            NSString *createTableFormat = @"CREATE TABLE IF NOT EXISTS %@ (";
            NSString *createLtsTriggerFormat = @"CREATE TRIGGER IF NOT EXISTS %@_check_lts BEFORE UPDATE OF lts ON %@ FOR EACH ROW WHEN OLD.deviceTs > OLD.lts BEGIN SELECT RAISE(ABORT, 'ignored') WHERE OLD.deviceTs <> NEW.lts; END";
            
            NSString *createFantomTriggerFormat = @"CREATE TRIGGER IF NOT EXISTS %@_fantom_%@ BEFORE INSERT ON %@ FOR EACH ROW WHEN NEW.%@ is not null BEGIN INSERT INTO %@ (id, isFantom, lts, deviceTs) SELECT NEW.%@, 1, null, null WHERE NOT EXISTS (SELECT * FROM %@ WHERE id = NEW.%@); END";
            NSString *updateFantomTriggerFormat = @"CREATE TRIGGER IF NOT EXISTS %@_fantom_%@_update BEFORE UPDATE OF %@ ON %@ FOR EACH ROW WHEN NEW.%@ is not null BEGIN INSERT INTO %@ (id, isFantom, lts, deviceTs) SELECT NEW.%@, 1, null, null WHERE NOT EXISTS (SELECT * FROM %@ WHERE id = NEW.%@); END";
            NSString *fantomIndexFormat = @"CREATE INDEX IF NOT EXISTS %@_isFantom on %@ (isFantom);";
            
            NSString *createCascadeTriggerFormat = @"DROP TRIGGER IF EXISTS %@_cascade_%@; CREATE TRIGGER IF NOT EXISTS %@_cascade_%@ BEFORE DELETE ON %@ FOR EACH ROW BEGIN DELETE FROM %@ WHERE %@ = OLD.id; END";
            
            NSString *isRemovedTriggerFormat = @"CREATE TRIGGER IF NOT EXISTS %@_isRemoved BEFORE INSERT ON %@ FOR EACH ROW BEGIN SELECT RAISE(IGNORE) FROM RecordStatus WHERE isRemoved = 1 AND objectXid = NEW.id LIMIT 1; END";
            
            
            for (NSString* entityName in entityNames){
                
                if ([entityName isEqualToString:@"STMSetting"] || [entityName isEqualToString:@"STMEntity"] ){
                    continue;
                }
                
                NSString *tableName = [self entityToTableName:entityName];
                NSMutableArray *columns = @[].mutableCopy;
                NSString *sql_stmt = [NSString stringWithFormat:createTableFormat, tableName];
                
                BOOL first = true;
                
                for (NSString* columnName in [STMCoreObjectsController allObjectsWithTypeForEntityName:entityName].allKeys){
                    
                    if ([columnName isEqualToString:@"xid"]) continue;
                    
                    if (first){
                        first = false;
                    }else{
                        sql_stmt = [sql_stmt stringByAppendingString:@", "];
                    }
                    
                    [columns addObject:columnName];
                    sql_stmt = [sql_stmt stringByAppendingString:columnName];
                    
                    NSAttributeDescription* atribute= [STMCoreObjectsController allObjectsWithTypeForEntityName:entityName][columnName];
                    
                    if ([columnName isEqualToString:@"id"]){
                        sql_stmt = [sql_stmt stringByAppendingString:@" TEXT PRIMARY KEY"];
                        continue;
                    }
                    
                    switch (atribute.attributeType) {
                        case NSStringAttributeType:
                        case NSDateAttributeType:
                        case NSUndefinedAttributeType:
                        case NSBinaryDataAttributeType:
                        case NSTransformableAttributeType:
                            sql_stmt = [sql_stmt stringByAppendingString:@" TEXT"];
                            break;
                        case NSInteger64AttributeType:
                        case NSBooleanAttributeType:
                        case NSObjectIDAttributeType:
                        case NSInteger16AttributeType:
                        case NSInteger32AttributeType:
                            sql_stmt = [sql_stmt stringByAppendingString:@" INTEGER"];
                            break;
                        case NSDecimalAttributeType:
                        case NSFloatAttributeType:
                        case NSDoubleAttributeType:
                            sql_stmt = [sql_stmt stringByAppendingString:@" NUMERIC"];
                            break;
                        default:
                            break;
                    }
                    
                    if ([columnName isEqualToString:@"deviceCts"]) {
                        sql_stmt = [sql_stmt stringByAppendingString:@" DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW'))"];
                    }
                    if ([columnName isEqualToString:@"lts"]) {
                        sql_stmt = [sql_stmt stringByAppendingString:@" DEFAULT('')"];
                    }
                }
                
                NSDictionary *relationships = [STMCoreObjectsController toOneRelationshipsForEntityName:entityName];
                
                for (NSString* entityKey in relationships.allKeys){
                    if (first){
                        first = false;
                    }else{
                        sql_stmt = [sql_stmt stringByAppendingString:@", "];
                    }
                    
                    NSString *fkColumn = [entityKey stringByAppendingString:RELATIONSHIP_SUFFIX];
                    NSString *fkTable = [self entityToTableName:[STMCoreObjectsController toOneRelationshipsForEntityName:entityName][entityKey]];
                    
                    NSString *cascadeAction = @"SET NULL";
                    NSString *fkSQL = [NSString stringWithFormat:fkColFormat, fkColumn, fkTable, cascadeAction];
                    
                    [columns addObject:fkColumn];
                    sql_stmt = [sql_stmt stringByAppendingString:fkSQL];
                }
                
                sql_stmt = [sql_stmt stringByAppendingString:@" ); "];
                columnsDictionary[[self entityToTableName:entityName]] = columns.copy;
 
                BOOL res = [database executeStatements:sql_stmt];
                NSLog(@"%@ (%@)",sql_stmt, res ? @"YES" : @"NO");
                
                sql_stmt = [NSString stringWithFormat:fantomIndexFormat, tableName, tableName];
                res = [database executeStatements:sql_stmt];
                NSLog(@"%@ (%@)",sql_stmt, res ? @"YES" : @"NO");

                sql_stmt = [NSString stringWithFormat:createLtsTriggerFormat, tableName, tableName, tableName];
                
                res = [database executeStatements:sql_stmt];
                NSLog(@"%@ (%@)", sql_stmt, res ? @"YES" : @"NO");
                
                sql_stmt = [NSString stringWithFormat:isRemovedTriggerFormat, tableName, tableName];
                
                res = [database executeStatements:sql_stmt];
                NSLog(@"%@ (%@)", sql_stmt, res ? @"YES" : @"NO");
                
                
                for (NSString* entityKey in [STMCoreObjectsController toOneRelationshipsForEntityName:entityName].allKeys){
                    NSString *fkColumn = [entityKey stringByAppendingString:RELATIONSHIP_SUFFIX];
                    
                    NSString *createIndexSQL = [NSString stringWithFormat:createIndexFormat, tableName, entityKey, tableName, fkColumn];
                    res = [database executeStatements:createIndexSQL];
                    NSLog(@"%@ (%@)", createIndexSQL, res ? @"YES" : @"NO");
                    
                    NSString *fkTable = [self entityToTableName:[STMCoreObjectsController toOneRelationshipsForEntityName:entityName][entityKey]];
                    
                    sql_stmt = [NSString stringWithFormat:createFantomTriggerFormat, tableName, fkColumn, tableName, fkColumn, fkTable, fkColumn, fkTable, fkColumn];
                    res = [database executeStatements:sql_stmt];
                    NSLog(@"%@ (%@)", sql_stmt, res ? @"YES" : @"NO");
                    
                    sql_stmt = [NSString stringWithFormat:updateFantomTriggerFormat, tableName, fkColumn, fkColumn, tableName, fkColumn, fkTable, fkColumn, fkTable, fkColumn];
                    res = [database executeStatements:sql_stmt];
                    NSLog(@"%@ (%@)", sql_stmt, res ? @"YES" : @"NO");
                    
                }
                
                NSDictionary <NSString *, NSRelationshipDescription*> *cascadeRelations = [STMCoreObjectsController objectRelationshipsForEntityName:entityName isToMany:@(YES) cascade:true];
                
                for (NSString* relationKey in cascadeRelations.allKeys){
                    
                    NSRelationshipDescription *relation = cascadeRelations[relationKey];
                    NSString *childTableName = [self entityToTableName:relation.destinationEntity.name];
                    NSString *fkColumn = [relation.inverseRelationship.name stringByAppendingString:@"Id"];
                    
                    sql_stmt = [NSString stringWithFormat:createCascadeTriggerFormat, tableName, relationKey,tableName, relationKey,tableName, childTableName, fkColumn];
                    res = [database executeStatements:sql_stmt];
                    NSLog(@"%@ (%@)", sql_stmt, res ? @"YES" : @"NO");
                    
                }
                
            }
            columnsByTable = columnsDictionary.copy;
        
        }];
    }
    return self;
}

- (NSString *)entityToTableName:(NSString *)entity{
    if ([entity hasPrefix:@"STM"]){
        return [entity substringFromIndex:3];
    }
    return entity ;
}

+ (STMFmdb *)sharedInstance {
    
    static dispatch_once_t pred = 0;
    __strong static id _sharedInstance = nil;
    
    dispatch_once(&pred, ^{
        _sharedInstance = [[self alloc] init];
    });
    
    return _sharedInstance;
    
}

- (NSDictionary * _Nullable)mergeIntoAndResponse:(NSString * _Nonnull)tablename dictionary:(NSDictionary<NSString *, id> * _Nonnull)dictionary error:(NSError *_Nonnull * _Nonnull)error{
    
    __block NSDictionary *response;
    
    [queue inDatabase:^(FMDatabase *db) {
        
        NSString *pk = [self mergeInto:tablename dictionary:dictionary error:error db:db];
        
        if (!pk) return;
        
        NSArray *results = [self getDataWithEntityName:tablename
                                         withPredicate:[NSPredicate predicateWithFormat:@"id == %@", pk]
                                               orderBy:nil
                                            fetchLimit:0
                                           fetchOffset:0
                                                    db:db];
        
        response = [results firstObject];
        
    }];
    
    return response;
    
}

- (BOOL) mergeInto:(NSString * _Nonnull)tablename dictionary:(NSDictionary<NSString *, id> * _Nonnull)dictionary error:(NSError *_Nonnull * _Nonnull)error {
    
    __block BOOL result;
    
    [queue inDatabase:^(FMDatabase *db) {
        result = !![self mergeInto:tablename
                        dictionary:dictionary
                             error:error
                                db:db];
    }];
    
    return result;
}

- (NSString *) mergeInto:(NSString * _Nonnull)tablename dictionary:(NSDictionary<NSString *, id> * _Nonnull)dictionary error:(NSError *_Nonnull * _Nonnull)error db:(FMDatabase *)db{
    
    tablename = [self entityToTableName:tablename];
    
    NSArray *columns = columnsByTable[tablename];
    NSString *pk = dictionary [@"id"] ? dictionary [@"id"] : [[[NSUUID alloc] init].UUIDString lowercaseString];
    
    NSMutableArray* keys = @[].mutableCopy;
    NSMutableArray* values = @[].mutableCopy;
    
    for(NSString* key in dictionary){
        if ([columns containsObject:key] && ![@[@"id", @"isFantom"] containsObject:key]){
            [keys addObject:key];
            id value = [dictionary objectForKey:key];
            if ([value isKindOfClass:[NSDate class]]) {
                [values addObject:[STMFunctions stringFromDate:(NSDate *)value]];
            } else {
                [values addObject:(NSString*)value];
            }
            
        }
    }
    
    [values addObject:pk];
    
    NSMutableArray* v = @[].mutableCopy;
    for (int i=0;i<[keys count];i++){
        [v addObject:@"?"];
    }
    
    NSString* updateSQL = [NSString stringWithFormat:@"UPDATE %@ SET isFantom = 0, %@ = ? WHERE id = ?", tablename, [keys componentsJoinedByString:@" = ?, "]];
    
    if(![db executeUpdate:updateSQL values:values error:error]){
        if ([[*error localizedDescription] isEqualToString:@"ignored"]){
            *error = nil;
            return pk;
        } else{
            return nil;
        }
    }
    
    if (!db.changes) {
        NSString *insertSQL = [NSString stringWithFormat:@"INSERT INTO %@ (%@, isFantom, id) VALUES(%@, 0, ?)", tablename, [keys componentsJoinedByString:@", "], [v componentsJoinedByString:@", "]];
        if (![db executeUpdate:insertSQL values:values error:error]){
            return nil;
        }
    }

    return pk;
}

- (NSArray * _Nonnull)getDataWithEntityName:(NSString * _Nonnull)name withPredicate:(NSPredicate * _Nonnull)predicate orderBy:(NSString * _Nullable)orderBy fetchLimit:(NSUInteger)fetchLimit fetchOffset:(NSUInteger)fetchOffset{
    
    __block NSArray* results;
    
    [pool inDatabase:^(FMDatabase *db) {
    
        results = [self getDataWithEntityName:name
                                withPredicate:predicate
                                      orderBy:orderBy
                                   fetchLimit:fetchLimit
                                  fetchOffset:fetchOffset
                                           db:db];
    
    }];
    
    return results;
}

<<<<<<< HEAD
- (NSArray * _Nonnull)getDataWithEntityName:(NSString * _Nonnull)name withPredicate:(NSPredicate * _Nonnull)predicate orderBy:(NSString * _Nullable)orderBy fetchLimit:(NSUInteger)fetchLimit fetchOffset:(NSUInteger)fetchOffset db:(FMDatabase *)db {
=======
- (BOOL)destroy:(NSString * _Nonnull)tablename identifier:(NSString*  _Nonnull)idendifier error:(NSError *_Nonnull * _Nonnull)error{
    
    __block BOOL result = YES;
    
    NSString* destroySQL = [NSString stringWithFormat:@"DELETE FROM %@ WHERE id=?", [self entityToTableName:tablename]];
    
    NSArray* values = @[idendifier];
    
    [queue inDatabase:^(FMDatabase *db) {
        if(![db executeUpdate:destroySQL values:values error:error]){
            result = NO;
        }
    }];
    
    return result;
}

- (NSArray * _Nonnull)getDataWithEntityName:(NSString * _Nonnull)name withPredicate:(NSPredicate * _Nonnull)predicate orderBy:(NSString * _Nullable)orderBy fetchLimit:(NSUInteger * _Nullable)fetchLimit fetchOffset:(NSUInteger * _Nullable)fetchOffset db:(FMDatabase *)db{
>>>>>>> persisting
    
    NSString* options = @"";
    
    if (orderBy) {
        NSString *order = [NSString stringWithFormat:@" ORDER BY %@", orderBy];
        options = [options stringByAppendingString:order];
    }
    
    if (fetchLimit) {
        NSString *limit = [NSString stringWithFormat:@" LIMIT %@", @(fetchLimit)];
        options = [options stringByAppendingString:limit];
    }
    
    if (fetchOffset) {
        NSString *offset = [NSString stringWithFormat:@" OFFSET %@", @(fetchOffset)];
        options = [options stringByAppendingString:offset];
    }
    
    name = [self entityToTableName:name];
    
    NSString* where = @"";
    
    if (predicate){
        where = [[STMPredicateToSQL sharedInstance] SQLFilterForPredicate:predicate entityName:name];
        if ([where isEqualToString:@"( )"] || [where isEqualToString:@"()"]){
            where = @"";
        }else{
            where = [@" WHERE " stringByAppendingString:where];
        }
    }
    
    where = [where stringByReplacingOccurrencesOfString:@" AND ()"
                                             withString:@""];
    where = [where stringByReplacingOccurrencesOfString:@"?uncapitalizedTableName?"
                                             withString:[STMFunctions lowercaseFirst:name]];
    where = [where stringByReplacingOccurrencesOfString:@"?capitalizedTableName?"
                                             withString:name];
    
    NSMutableArray *rez = @[].mutableCopy;

    NSString* query = [NSString stringWithFormat:@"SELECT * FROM %@%@%@", name, where, options];
    
    FMResultSet *s = [db executeQuery:query];
    
    while ([s next]) {
        [rez addObject:s.resultDictionary];
    }
    
    // there will be memory warnings loading catalogue on an old device if no copy
    return rez.copy;

}

- (BOOL) hasTable:(NSString * _Nonnull)name {
    name = [self entityToTableName:name];
    if ([columnsByTable.allKeys containsObject:name]){
        return true;
    }
    return false;
}

- (NSArray * _Nonnull) allKeysForObject:(NSString * _Nonnull)obj {
    obj = [self entityToTableName:obj];
    return columnsByTable[obj];
}

- (BOOL) commit {
    __block BOOL result = YES;
    [queue inDatabase:^(FMDatabase *db) {
        if ([db inTransaction]){
            result = [db commit];
        }
    }];
    return result;
}

- (BOOL) startTransaction {
    __block BOOL result = YES;
    [queue inDatabase:^(FMDatabase *db){
        if (![db inTransaction]){
            result = [db beginTransaction];
        }
    }];
    return result;
}
     
- (BOOL) rollback {
    __block BOOL result = YES;
    [queue inDatabase:^(FMDatabase *db){
        [db rollback];
    }];
    return result;
}

@end

File: iSistemiumCore/DataModel/STMFmdb.swift
//
//  STMFmdb.swift
//  iSisSales
//
//  Created by Edgar Jan Vuicik on 12/12/2016.
//  Copyright © 2016 Sistemium UAB. All rights reserved.
//

import FMDB

@objc
class STMFmdb:NSObject{
    
    static let sharedInstance = STMFmdb()
    
    let databaseName = "database.db"
    
    private var database:FMDatabase!
    
    private override init(){
        let filemgr = FileManager.default
        let dirPaths = filemgr.urls(for: .documentDirectory,
                                    in: .userDomainMask)
        
        let databasePath = dirPaths[0].appendingPathComponent(databaseName).path
        
        database = FMDatabase(path: databasePath as String)
        
        if database == nil {
            NSLog("STMFmdb error: \(database?.lastErrorMessage())")
        }
        
        if !filemgr.fileExists(atPath: databasePath as String) {
            //STMEntity
            NSLog("Started creating fmdb tables")
            if (database?.open())!{
                database.beginTransaction()
                var sql_stmt = "CREATE TABLE IF NOT EXISTS STMPrice (ID TEXT PRIMARY KEY, commentText TEXT, deviceCts TEXT, deviceTs TEXT, INTEGER, lts TEXT, ownerXid TEXT, source TEXT, target TEXT, price NUMERIC, articleid TEXT REFERENCES STMArticle(id), pricetypeid TEXT REFERENCES STMPriceType(id) ) "
                if !(database?.executeStatements(sql_stmt))! {
                    NSLog("STMFmdb error: \(database?.lastErrorMessage())")
                }
                sql_stmt = "CREATE TABLE IF NOT EXISTS STMArticle (ID TEXT PRIMARY KEY, commentText TEXT, deviceCts TEXT, deviceTs TEXT, INTEGER, lts TEXT, ownerXid TEXT, source TEXT, target TEXT, barcode TEXT, code TEXT, extraLabel TEXT,factor INTEGER, name TEXT, packageRel INTEGER, pieceVolume NUMERIC, pieceWeight NUMERIC, price NUMERIC, articlegroupid TEXT REFERENCES STMArticleGroup(id)) "
                if !(database?.executeStatements(sql_stmt))! {
                    NSLog("STMFmdb error: \(database?.lastErrorMessage())")
                }
                sql_stmt = "CREATE TABLE IF NOT EXISTS STMStock (ID TEXT PRIMARY KEY, commentText TEXT, deviceCts TEXT, deviceTs TEXT, INTEGER, lts TEXT, ownerXid TEXT, source TEXT, target TEXT, displayVolume TEXT, volume INTEGER, articleid TEXT REFERENCES STMArticle(id)) "
                if !(database?.executeStatements(sql_stmt))! {
                    NSLog("STMFmdb error: \(database?.lastErrorMessage())")
                }
                sql_stmt = "CREATE TABLE IF NOT EXISTS STMArticleGroup (ID TEXT PRIMARY KEY, commentText TEXT, deviceCts TEXT, deviceTs TEXT, INTEGER, lts TEXT, ownerXid TEXT, source TEXT, target TEXT, name TEXT, articlegroupid TEXT REFERENCES STMArticleGroup(id)) "
                if !(database?.executeStatements(sql_stmt))! {
                    NSLog("STMFmdb error: \(database?.lastErrorMessage())")
                }
//                sql_stmt = "CREATE TABLE IF NOT EXISTS STMPriceType (ID TEXT PRIMARY KEY, commentText TEXT, deviceCts TEXT, deviceTs TEXT, INTEGER, lts TEXT, ownerXid TEXT, source TEXT, target TEXT, name TEXT, parentId TEXT REFERENCES STMPriceType(id)) "
//                if !(database?.executeStatements(sql_stmt))! {
//                    NSLog("STMFmdb error: \(database?.lastErrorMessage())")
//                }
                sql_stmt = "CREATE TABLE IF NOT EXISTS STMSaleOrderPosition (ID TEXT PRIMARY KEY, commentText TEXT, deviceCts TEXT, deviceTs TEXT, INTEGER, lts TEXT, ownerXid TEXT, source TEXT, target TEXT, backVolume INTEGER, cost NUMERIC, price NUMERIC, priceDoc NUMERIC, priceOrigin NUMERIC, volume INTEGER, articleid TEXT REFERENCES STMArticle(id), saleorderId TEXT REFERENCES STMSaleOrder(id)) "
                
                if !(database?.executeStatements(sql_stmt))! {
                    NSLog("STMFmdb error: \(database?.lastErrorMessage())")
                }
                database?.commit();
                database?.close()
            } else {
                NSLog("STMFmdb error: \(database?.lastErrorMessage())")
            }
            NSLog("Done creating fmdb tables")
        }
    }
    
    func insert(tablename:String, array:Array<Dictionary<String, Any>>, completionHandler:(_ success:Bool)->Void){
        NSLog("Started inserting \(tablename)")
        if (database?.open())! {
            database.beginTransaction()
            for dictionary in array{
                insert(tablename: tablename, dictionary: dictionary)
            }
            database?.commit();
            database?.close()
            completionHandler(true);
        } else {
            NSLog("STMFmdb error: \(database?.lastErrorMessage())")
            completionHandler(false);
        }
        NSLog("Done inserting \(tablename)")
    }

    private func insert(tablename:String, dictionary:Dictionary<String, Any>){
        var keys:[String] = []
        
        var values:[Any] = []
        
        for (key, value) in dictionary{
            switch(key){
            case "ts":
                break
            case "discountPercent":
                break
            case "author":
                break
            case "articleSameId":
                break
            case "productionInfoType":
                break
            default:
                keys.append(key)
                values.append(value)
            }
        }
        
        keys.append("lts")
        values.append("'\(Date())'")
        let v = [String](repeating: "?", count: keys.count)
        let insertSQL = "INSERT OR REPLACE INTO \(tablename) (\(keys.joined(separator: ", "))) VALUES (\(v.joined(separator: ", ")))"
        
        let result = database?.executeUpdate(insertSQL,
                                             withArgumentsIn: values)
        
        if !result! {
            NSLog("STMFmdb error: \(database?.lastErrorMessage())")
        }
    }
    
    func insert(tablename:String, dictionary:Dictionary<String, Any>, completionHandler:(_ success:Bool)->Void){
        if (database?.open())! {
            insert(tablename: tablename, dictionary: dictionary)
            completionHandler(true);
        } else {
            NSLog("STMFmdb error: \(database?.lastErrorMessage())")
            completionHandler(false);
        }
    }
    
//    func select(tablename:String){
//    SELECT name FROM my_db.sqlite_master WHERE type='table';
//        if (database?.open())! {
//            let sql_stmt = "SELECT * FROM STMPrice CROSS JOIN STMArticle"
//            let rs = try? database.executeQuery(sql_stmt,values:[])
//            while rs!.next() {
//                rs?.resultDictionary().forEach { print("\($0): \($1)") }
//            }
//            database?.close()
//        } else {
//            NSLog("STMFmdb error: \(database?.lastErrorMessage())")
//        }
//    }
    
    func getTableNames()->[String]{
        var rez = [String]()
        if (database?.open())! {
            let sql_stmt = "SELECT name FROM sqlite_master WHERE type='table'"
            let rs = try? database.executeQuery(sql_stmt,values:[])
            while rs!.next() {
                rez.append(rs?.resultDictionary()["name"] as! String)
            }
            database?.close()
        } else {
            NSLog("STMFmdb error: \(database?.lastErrorMessage())")
        }
        return rez
    }
    
    func containstTableWithName(name:String)->Bool{
        return getTableNames().contains(name)
    }
    
}
